Modules/arraymodule.c

```diff
@@ -3090,11 +3090,16 @@ array_arrayiterator___setstate__(arrayiterobject *self, PyObject *state)
     Py_ssize_t index = PyLong_AsSsize_t(state);
     if (index == -1 && PyErr_Occurred())
         return NULL;
-    if (index < 0)
-        index = 0;
-    else if (index > Py_SIZE(self->ao))
-        index = Py_SIZE(self->ao); /* iterator exhausted */
-    self->index = index;
+    arrayobject *ao = self->ao;
```

Comment:
Can't you just check that before doing `PyLong_AsSsize_t`? Something like that:

```c
    if (self->ao == NULL) {
       goto exit;
    }
    Py_ssize_t index = ...;
    ...
exit:
    Py_RETURN_NONE;
}
```

Or if you think there's a possibility for `self->ao` to become NULL between the beginning of the function and the end, maybe something like

```c
    Py_ssize_t index = ...;
    ...
    if (self->ao == NULL) {
        goto exit;
    }
    if (index < 0) {
        index = 0;
    }
    ...
exit:
    Py_RETURN_NONE;
```

Would it be safe on the free-threaded build by the way?

Reply:
1. You probably want an erroneous `state` value to trigger an error regardless, no?
2. It is not free-threaded safe, which is why #128943 exists and this actually was pulled out from there standalone to be backported.
3. I was minimizing goto usage as per @ZeroIntensity request, but maybe I misunderstood him and he just meant don't use just in critical sections. Anyways, tell me how u want it and I will change if needed.

Reply:
Oh yes so since we need freethreaded safety afterwards, better to avoid gotos.  As for checking the state value before usage, it makes sense.